### ISSUE TITLE
Validation

### ISSUE THREAD

**Issue by detrin** [manuel-calzolari/sklearn-genetic#38]
I am playing around the repository and in order to have some kind of added value for my projects I think the project is missing one key thing. Validation on validation dataset. The `GeneticSelectionCV` will simply overfit on data and model on test sample that is taken apart will have nill performance. 

Instead of 
```python
selector = selector.fit(
    data.loc[train_mask, cols_pred],
    data.loc[train_mask, col_target]
)
```
The API should look something as follows
```python
selector = selector.fit(
    data.loc[train_mask, cols_pred],
    data.loc[train_mask, col_target],
    X_valid = data.loc[valid_mask, cols_pred],
    y_valid = data.loc[valid_mask, col_target]
)
```

What do you think?

**Comment by manuel-calzolari** [comment]
Hi @detrin, thank you for your feedback and your interest in this project.

The current design of `GeneticSelectionCV` is intended to adhere to the scikit-learn API as closely as possible, and components in scikit-learn like `RFECV` utilize an interface where the validation set is not explicitly specified in the `fit` method.

In terms of overfitting, `GeneticSelectionCV` (as well as `RFECV`) mitigates this concern by default through k-fold cross-validation. The idea behind k-fold cross-validation is that it leverages the entirety of your dataset for both training and validation at different iterations, which provides a more comprehensive assessment of the model's performance as compared to using a single hold-out validation set. The average score across all folds is generally a better estimate of the model performance.

However, I also understand that in some scenarios you may prefer using a hold-out validation set. Therefore, just as with `RFECV` and other scikit-learn components, you can use a hold-out validation set by providing an iterable that generates (train, test) splits as arrays of indices. This allows you to use a single, separate validation dataset if it better suits your project's requirements.

**Comment by detrin** [comment]
> The current design of GeneticSelectionCV is intended to adhere to the scikit-learn API as closely as possible, and components in scikit-learn like RFECV utilize an interface where the validation set is not explicitly specified in the fit method.
I respect this vision.

> However, I also understand that in some scenarios you may prefer using a hold-out validation set. Therefore, just as with RFECV and other scikit-learn components, you can use a hold-out validation set by providing an iterable that generates (train, test) splits as arrays of indices. This allows you to use a single, separate validation dataset if it better suits your project's requirements.
Could you provide please minimal viable example?

The issue that I am facing is maybe a little bit different. It is similar to hyperoptimization of parameters, except the hyperparamaters here are binary options on what features should be selected. Probably you are familiar with Stepwise Logistic Regression, not specifically using Logistic Regression model as base model, but generally regression. stepwise feature selection works only when there are negligible interactions between features, which in real world is almost never true. Therefore I am exploring other options for some time now. 

The problem is as follows. There is training set that could be split into training and valid set with CV as you described. However what is the one fit of the model is taking minutes already, then it is not beneficial to do CV if the validation set is sufficiently big (imagine that the amount of data is not a problem). 

Is the decision making of what features should be added or excluded made on validation part of CV using scoring function?

**Comment by detrin** [comment]
According to https://github.com/manuel-calzolari/sklearn-genetic/blob/master/genetic_selection/gscv.py#L126 you make decision based on validation part of dataset in CV.

**Comment by detrin** [comment]
There is also another thing I wanted to suggest, perhaps it is better for new issue. But I will mention it here as there are not many issues opened anyway. That is custom scoring function. What if I have scoring function that is not available in sklearn, but I use it for hyperoptimisation in feat selection? Do you think it would be possible and desirable to incorporate the ability to add custom scoring function into your package?

**Comment by manuel-calzolari** [comment]
Hi @detrin!

> Could you provide please minimal viable example?

You can use [PredefinedSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.PredefinedSplit.html) from sklearn. Example:
```
import numpy as np
from sklearn.model_selection import train_test_split, PredefinedSplit
from sklearn.ensemble import RandomForestClassifier
from genetic_selection import GeneticSelectionCV

...

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2)

fold = [-1] * len(X_train) + [0] * len(X_val)
ps = PredefinedSplit(fold)

X_combined = np.vstack((X_train, X_val))
y_combined = np.concatenate((y_train, y_val))

estimator = RandomForestClassifier()
selector = GeneticSelectionCV(estimator, cv=ps, scoring="accuracy")
selector = selector.fit(X_combined, y_combined)
```

> There is also another thing I wanted to suggest, perhaps it is better for new issue. But I will mention it here as there are not many issues opened anyway. That is custom scoring function. What if I have scoring function that is not available in sklearn, but I use it for hyperoptimisation in feat selection? Do you think it would be possible and desirable to incorporate the ability to add custom scoring function into your package?

This is already possible as with sklearn's feature selection modules using [make_scorer](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html). Example:
```
import numpy as np
from sklearn.model_selection import train_test_split, PredefinedSplit
from sklearn.ensemble import RandomForestClassifier
from genetic_selection import GeneticSelectionCV
from sklearn.metrics import make_scorer

...

def custom_score(y_true, y_pred):
    return np.mean(y_true == y_pred)

custom_scorer = make_scorer(custom_score, greater_is_better=True)

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2)

fold = [-1] * len(X_train) + [0] * len(X_val)
ps = PredefinedSplit(fold)

X_combined = np.vstack((X_train, X_val))
y_combined = np.concatenate((y_train, y_val))

estimator = RandomForestClassifier()
selector = GeneticSelectionCV(estimator, cv=ps, scoring=custom_scorer)
selector = selector.fit(X_combined, y_combined)
```

**Comment by detrin** [comment]
@manuel-calzolari  thank you, that is actually really helpful.

The class `PredefinedSplit` is obscure ina sense I have never seen it being used anywhere https://stackoverflow.com/a/56793912 However, this will provide a workaround for creating just train/valid split instead of CV as asked.
